pygatt/bled112_backend.py

```diff
@@ -0,0 +1,1977 @@
+            self._main_thread_cond.wait()
+        self._response_received = False  # reset the flag
+
+        # Start encryption
+        self._logger.info("encrypt_start")
+        cmd = self._lib.ble_cmd_sm_encrypt_start(
+            self._connection_handle, bonding['do_not_create_bonding'])
+        self._lib.send_command(self._ser, cmd)
+
+        # Wait for response
+        while not self._response_received:
+            self._main_thread_cond.wait()
+        self._response_received = False  # reset the flag
+        if self._response_return != 0:
+            self._loglock.acquire()
+            self._logger.warn("encrypt_start failed %s",
+                              get_return_message(self._response_return))
+            self._loglock.release()
+            self._main_thread_cond.release()
+            return
+
+        # Wait for event
+        while (not self._encrypted) and self._connected:
+            self._main_thread_cond.wait()
+        self._loglock.acquire()
+        if not self._connected:
+            self._logger.warn("encrypt_start failed: disconnected")
+        elif self._bonding_fail:
+            # Device may try to bond and cause this, so catch if possible
+            self._bonding_fail = False
+
+        # Drop locks
+        self._loglock.release()
+        self._main_thread_cond.release()
+
+    def get_devices_discovered(self):
+        """
+        Get self._devices_discovered in a thread-safe way.
+        A scan() should be run prior to accessing this data.
+
+        Returns the self._devices_discovered dictionary.
+        """
+        # get locks
+        self._main_thread_cond.acquire()
+        self._loglock.acquire()
+
+        # Log
+        self._logger.info("get_devices_discovered")
+
+        devs = self._devices_discovered
+
+        # Drop locks
+        self._loglock.release()
+        self._main_thread_cond.release()
+
+        return devs
+
+    def get_handle(self, characteristic_uuid, descriptor_uuid=None):
+        """
+        Get the handle for a characteristic or descriptor.
+
+        This requires that a connection is already established with the device.
+
+        characteristic_uuid -- bytearray containing the characteristic UUID.
+        descriptor_uuid -- optional bytearray containg the GATT descriptor UUID
+                           for the given characteristic. Note: use the
+                           gatt_characteristic_descriptor_uuid constant.
+
+        Returns an integer containing the handle on success.
+        Returns None on failure.
+        """
+        # get locks
+        self._main_thread_cond.acquire()
+        self._loglock.acquire()
+
+        # Make sure there is a connection
+        if not self._connected:
+            self._logger.warn("Not connected")
+            self._loglock.release()
+            self._main_thread_cond.release()
+            return
+
+        # Discover characteristics if not cached
+        if not self._characteristics_cached:
```

Comment:
Are you talking aobut characteristics_cached? It acts like a static variable in a C function. Since python doesn't have static variables and it needs to be persistent across calls to get_handle, I don't think so.